### PR TITLE
Lookup Joins

### LINKED ISSUE #3153: Lookup Joins
https://github.com/cockroachdb/cockroach/issues/19038
https://github.com/cockroachdb/cockroach/issues/25431
https://github.com/cockroachdb/cockroach/pull/22674

explain in docs that there are multiple join algorithms. Explain how each algorithm gets chosen, and how users can influence that choice. (There is a little mention of algorithms in the perf best practices, but the primary source of information should be joins.md instead.) - #534

### PR DESCRIPTION
**Pull request by lhirata**
- Added info about lookup joins + consolidated info about join algorithms on the Join Expression page.
- Cleaned up style on headings and mentions of joins.

Closes #3153.

### CODE REVIEW

**Comment by jordanlewis** [comment]
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[v2.1/joins.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsCQvo8CWCMQgOcYHU:-LDsCQvo8CWCMQgOcYHV:bjgmr2r) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L105)):*
> ```Markdown
> ### Merge joins
> 
> By default, CockroachDB uses [merge joins](https://en.wikipedia.org/wiki/Sort-merge_join) whenever possible because they are more performant than [hash joins](#hash-joins), computationally and in terms of memory. A merge join requires both tables to be indexed on the equality columns; when this condition is not met, CockroachDB resorts to the slower hash joins. Merge joins can be used only with [distributed query processing](https://www.cockroachlabs.com/blog/local-and-distributed-processing-in-cockroachdb/).
> ```

The condition is actually slightly more complicated - merge joins can only occur when the equality columns are indexed *and* the indexes both have the same ordering. That last bit is important.

---

*[v2.1/joins.md, line 109 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsD9xl1ljBoOJ2FchQ:-LDsD9xl1ljBoOJ2FchR:b-ds5pqt) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L109)):*
> ```Markdown
> Merge joins are performed on the indexed columns of two tables as follows:
> 
> 1. CockroachDB checks for indexes on the equality columns and that they are ordered the same (i.e., `ASC` or `DESC`).
> ```

Ah I see that you mention the ordering requirement down here. You should put it above as well.

---

*[v2.1/joins.md, line 111 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDHSE64IRAEf1vEve:-LDsDHSE64IRAEf1vEvf:b-i3hroj) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L111)):*
> ```Markdown
> 1. CockroachDB checks for indexes on the equality columns and that they are ordered the same (i.e., `ASC` or `DESC`).
> 2. CockroachDB takes one row from each table and compares them.
>     - If the rows are equal, CockroachDB returns the rows.
> ```

Unfortunately it's a bit more complicated than this. The semantics depend on the join type as well (inner/outer), and more stuff has to happen if there are multiple matches - the cartesian product of the matches is emitted.

---

*[v2.1/joins.md, line 121 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDk4O4aodyGj3AckP:-LDsDk4O4aodyGj3AckQ:ba3y1h4) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L121)):*
> ```Markdown
> 
> 1. CockroachDB reads both tables and attempts to pick the smaller table.
> 2. CockroachDB creates an in-memory [hash table](https://en.wikipedia.org/wiki/Hash_table) on the smaller table.
> ```

You might want to mention that this stage can fall back to disk if the hash table grows too large.

---

*[v2.1/joins.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDpjx5GIh_aHu0i_Q:-LDsDpjx5GIh_aHu0i_R:b-olq06v) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L122)):*
> ```Markdown
> 1. CockroachDB reads both tables and attempts to pick the smaller table.
> 2. CockroachDB creates an in-memory [hash table](https://en.wikipedia.org/wiki/Hash_table) on the smaller table.
> 3. CockroachDB then uses the hash table and scans the larger table to find matching rows from the smaller table.
> ```

maybe reword this: `CockroachDB then scans the larger table, looking up each row in the hash table.`

---

*[v2.1/joins.md, line 128 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDyog1g-5893otA5S:-LDsDyog1g-5893otA5T:bm2tsh5) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L128)):*
> ```Markdown
> {% include experimental-warning.md %}
> 
> Lookup joins are beneficial to use when there is a large imbalance in size between the two tables, as it only reads the smaller table and then looks up matches in the larger table. A lookup join requires both tables to be indexed on the equality columns.
> ```

singular/plural mismatch - "Lookup joins" vs "as it only reads"

The requirement isn't quite right - with lookup joins, only the right-hand table needs to be indexed on the equality columns.

---

*[v2.1/joins.md, line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsElPf2a2MZnT_7CGv:-LDsElPf2a2MZnT_7CGw:bvip0vd) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L141)):*
> ```Markdown
> 2. Specify the indexes to use if not the default index.
> 4. CockroachDB reads each row in the small table.
> 5. CockroachDB then scans (or "looks up") the larger table for matches to the smaller table and outputs the matching rows.
> ```

This list of steps feels heterogeneous. The first two steps are things the user does. The third and fourth are things the database does.

I would separate this into two lists: how to use the join, and how the database performs them.

In the first list, you should include an example of an actual join statement to demonstrate what you mean by "specify the indexes", and to show that there's no special join syntax to be used - it's all handled by that `SET` flag.

---

*[v2.1/performance-best-practices-overview.md, line 153 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsFEr1-1D0PazdqF_t:-LDsFEr1-1D0PazdqF_u:b-3c3ikh) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/performance-best-practices-overview.md#L153)):*
> ```Markdown
> ### Use indexes for faster joins
> 
> See [Join Performance Best Practices](joins.html#performance-best-practices).
> ```

Oops, I didn't realize that you were just moving the text about joins from a different spot. In that case, feel free to ignore my comments about the join descriptions (although we should fix them eventually, I think).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226)*

**Comment by lhirata** [comment]
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[v2.1/joins.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsCQvo8CWCMQgOcYHU:-LDwb9mH8ZbWgsAbjP71:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L105)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

The condition is actually slightly more complicated - merge joins can only occur when the equality columns are indexed *and* the indexes both have the same ordering. That last bit is important.
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 109 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsD9xl1ljBoOJ2FchQ:-LDwbB_75alWVqJRtX1F:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L109)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ah I see that you mention the ordering requirement down here. You should put it above as well.
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 111 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDHSE64IRAEf1vEve:-LDxOQonCCqdVr2tppJa:b-xci1nc) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L111)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Unfortunately it's a bit more complicated than this. The semantics depend on the join type as well (inner/outer), and more stuff has to happen if there are multiple matches - the cartesian product of the matches is emitted.
</blockquote></details>

Tried to clarify - is this better?

---

*[v2.1/joins.md, line 121 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDk4O4aodyGj3AckP:-LDwbux3DnvxYCbIzRpu:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L121)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You might want to mention that this stage can fall back to disk if the hash table grows too large.
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDpjx5GIh_aHu0i_Q:-LDwc24I7W4D81BXwtpk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L122)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

maybe reword this: `CockroachDB then scans the larger table, looking up each row in the hash table.`
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 128 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDyog1g-5893otA5S:-LDwcRMcBPr6m__jKNJ5:bze7x1p) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L128)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

singular/plural mismatch - "Lookup joins" vs "as it only reads"

The requirement isn't quite right - with lookup joins, only the right-hand table needs to be indexed on the equality columns.
</blockquote></details>

Right-hand meaning the larger table?

---

*[v2.1/joins.md, line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsElPf2a2MZnT_7CGv:-LDxEgQl02LPX79kEKEP:bg0qjkx) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L141)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This list of steps feels heterogeneous. The first two steps are things the user does. The third and fourth are things the database does.

I would separate this into two lists: how to use the join, and how the database performs them.

In the first list, you should include an example of an actual join statement to demonstrate what you mean by "specify the indexes", and to show that there's no special join syntax to be used - it's all handled by that `SET` flag.
</blockquote></details>

Is this syntax correct?

---

*[v2.1/performance-best-practices-overview.md, line 153 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsFEr1-1D0PazdqF_t:-LDwcaa4-wseDzy45AKL:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/performance-best-practices-overview.md#L153)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Oops, I didn't realize that you were just moving the text about joins from a different spot. In that case, feel free to ignore my comments about the join descriptions (although we should fix them eventually, I think).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226)*

**Comment by jordanlewis** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[v2.1/joins.md, line 111 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDHSE64IRAEf1vEve:-LE1GscNE4wjhtjs2x8x:b9bz4sq) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L111)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Tried to clarify - is this better?
</blockquote></details>

Yeah, I think so. It's still not a full and unambiguous description of the algorithm, but that's probably not necessary for these docs.

---

*[v2.1/joins.md, line 128 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDyog1g-5893otA5S:-LE1H1e97OygR_fEs0ko:bb77ufg) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L128)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Right-hand meaning the larger table?
</blockquote></details>

Yes.

---

*[v2.1/joins.md, line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsElPf2a2MZnT_7CGv:-LE1HK9R3J8wP2DEelOP:b-idzi11) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L141)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Is this syntax correct?
</blockquote></details>

It should be `SELECT * FROM weather@index_1`. Also, and I forgot to mention this, this isn't valid - lookup joins are only valid on inner or left outer joins.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226)*

**Comment by lhirata** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[v2.1/joins.md, line 111 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDHSE64IRAEf1vEve:-LE1OgRd2kfaOONfPPYi:b-61fku) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L111)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yeah, I think so. It's still not a full and unambiguous description of the algorithm, but that's probably not necessary for these docs.
</blockquote></details>

Ok, I can revisit this and try to iterate on it later, but I think it's ok for the alpha.

---

*[v2.1/joins.md, line 128 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsDyog1g-5893otA5S:-LE1POn3-o8QmV7qs-tI:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L128)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yes.
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/3226#-LDsElPf2a2MZnT_7CGv:-LE1Pfui3uidgFIFcqfB:bnhidst) ([raw file](https://github.com/cockroachdb/docs/blob/c91d2c1921ca422815db2df1dfb23527d280d843/v2.1/joins.md#L141)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It should be `SELECT * FROM weather@index_1`. Also, and I forgot to mention this, this isn't valid - lookup joins are only valid on inner or left outer joins.
</blockquote></details>

Updated. Added a note about lookup joins only being valid for inner/left outer joins.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226)*

**Comment by lhirata** [comment]
TFTR @jordanlewis! 

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226#-:-LE1Pv_fE5H4JsvREDPX:b7nmb66)*

**Comment by jordanlewis** [comment]
LGTM

**Comment by jseldess** [comment]
:lgtm:, with some comments.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[v2.0/create-view.md, line 31 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9wa3oDXerJCujR83f:-LE9wa3oDXerJCujR83g:b9191e0) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.0/create-view.md#L31)):*
> ```Markdown
> {{site.data.alerts.callout_success}}This example highlights one key benefit to using views: simplifying complex queries. For additional benefits and examples, see <a href="views.html">Views</a>.{{site.data.alerts.end}}
> 
> Let's say you're using our [sample `startrek` database](generate-cockroachdb-resources.html#generate-example-data), which contains two tables, `episodes` and `quotes`. There's a foreign key constraint between the `episodes.id` column and the `quotes.episode` column. To count the number of famous quotes per season, you could run the following join:
> ```

Do you want me to add this to the https://github.com/cockroachdb/docs/blob/master/ci/vale/CockroachDB/Terminology.yml rule to make sure we're consistent about "join" instead of "`JOIN`" across our docs?

---

*[v2.1/joins.md, line 118 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9tuUc9YnguavsnbfP:-LE9tuUc9YnguavsnbfQ:b9lqudv) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L118)):*
> ```Markdown
> 
> 
>     For outer joins:
> ```

For some reason, this second nested list isn't rendering properly:



You might try remove the empty lines between the last bullet of the previous list and `For outer joins`.

---

*[v2.1/joins.md, line 139 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9ubut-G7vP281iFY9:-LE9ubut-G7vP281iFYA:bxe34jp) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L139)):*
> ```Markdown
> A lookup join is beneficial to use when there is a large imbalance in size between the two tables, as it only reads the smaller table and then looks up matches in the larger table. A lookup join requires that the right-hand (i.e., larger) table is indexed on the equality column.
> 
> {{site.data.alerts.callout_info}}Lookup joins are only valid on inner or left outer joins.{{site.data.alerts.end}}
> ```

For clarity, I'd say: `only valid on inner joins and left outer joins.`

---

*[v2.1/joins.md, line 143 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9uoXX5XqnMLbL_loC:-LE9uoXX5XqnMLbL_loD:b2neypc) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L143)):*
> ```Markdown
> To use a lookup join:
> 
> 1. Manually set the flag:
> ```

It might help some users to explicitly reference our built-in sql client, e.g.:

```
1. Open the [built-in SQL shell](use-the-built-in-sql-client.html) and enable the feature:
```

---

*[v2.1/joins.md, line 150 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9vT9O-MB8We_sobEF:-LE9vT9O-MB8We_sobEG:b-szcjwg) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L150)):*
> ```Markdown
>     ~~~
> 
> 2. Specify the indexes to use if not the default index. For example:
> ```

nit: `In your query, specify the indexes...`

---

*[v2.1/joins.md, line 168 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9vn_923urPxSrwIER:-LE9vn_923urPxSrwIES:b-c600z4) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L168)):*
> ```Markdown
> - Joins over [interleaved tables](interleave-in-parent.html) are usually (but not always) processed more effectively than over non-interleaved tables.
> - When no indexes can be used to satisfy a join, CockroachDB may load all the rows in memory that satisfy the condition one of the join operands before starting to return result rows. This may cause joins to fail if the join condition or other `WHERE` clauses are insufficiently selective.
> - Outer joins are generally processed less efficiently than inner joins. Prefer using inner joins whenever possible. Full outer joins are the least optimized.
> ```

It'd be helpful to make `Outer joins` and `inner joins` links to the relevant sections on this page, especially for users jumping here straight from the performance optimization best practices page. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226#-:-LE9wupD0YH_sBK-ylg4:b-64wdml)*

**Comment by rmloveland** [comment]
:lgtm:

---

Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[v2.1/joins.md, line 119 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LEA1ib_49-m5VFVmeii:-LEA1ib_49-m5VFVmeij:b-jgigon) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L119)):*
> ```Markdown
> 
>     For outer joins:
>     - If the rows are equal, CockroachDB returns the rows.
> ```

Not really relevant for this PR, but wherever we describe algorithms in text I would love to (usually) also have diagrams to show them.  Since the reader has to build a picture in their mind from the text anyway, we might be able to save them some time/effort.  Will file an issue if none exist.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226#-:-LEA4-_D2zNHjGn0EFJo:bnfp4nl)*

**Comment by lhirata** [comment]
Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[v2.0/create-view.md, line 31 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9wa3oDXerJCujR83f:-LEAHLnREygTmm_nroVB:bum6hwh) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.0/create-view.md#L31)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Do you want me to add this to the https://github.com/cockroachdb/docs/blob/master/ci/vale/CockroachDB/Terminology.yml rule to make sure we're consistent about "join" instead of "`JOIN`" across our docs?
</blockquote></details>

Yeah, I think that would be good. Can we ignore `JOIN` when it's used in a SQL statement?

---

*[v2.1/joins.md, line 118 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9tuUc9YnguavsnbfP:-LEAJ19z8WN7BBRO2n8g:b-538mc6) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L118)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

For some reason, this second nested list isn't rendering properly:



You might try remove the empty lines between the last bullet of the previous list and `For outer joins`.
</blockquote></details>

The formatting is being finicky, so I just made "For inner joins:" and "For outer joins" bullets as well.

---

*[v2.1/joins.md, line 119 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LEA1ib_49-m5VFVmeii:-LEAJSIL7STDTgjJY3w6:b-o21qkp) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L119)):*
<details><summary><i>Previously, rmloveland (Rich Loveland) wrote…</i></summary><blockquote>

Not really relevant for this PR, but wherever we describe algorithms in text I would love to (usually) also have diagrams to show them.  Since the reader has to build a picture in their mind from the text anyway, we might be able to save them some time/effort.  Will file an issue if none exist.
</blockquote></details>

Good idea! I don't think there's an issue (at least for joins). I can file an issue for joins, but you might want to add another ticket if there are other algorithms that we should diagram.

---

*[v2.1/joins.md, line 139 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9ubut-G7vP281iFY9:-LEAK8sK-sj5QYrN8fgF:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L139)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

For clarity, I'd say: `only valid on inner joins and left outer joins.`
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 143 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9uoXX5XqnMLbL_loC:-LEAKGdO00jeouqLA8fb:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L143)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

It might help some users to explicitly reference our built-in sql client, e.g.:

```
1. Open the [built-in SQL shell](use-the-built-in-sql-client.html) and enable the feature:
```
</blockquote></details>

Done.

---

*[v2.1/joins.md, line 168 at r3](https://reviewable.io/reviews/cockroachdb/docs/3226#-LE9vn_923urPxSrwIER:-LEAL4C83mJEslbEn97D:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/86853a0ebac57f9f9901b3101d24241e596d34b9/v2.1/joins.md#L168)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

It'd be helpful to make `Outer joins` and `inner joins` links to the relevant sections on this page, especially for users jumping here straight from the performance optimization best practices page. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3226)*
